File: Tests/Tests/STPAPIClientNetworkBridgeTest.m
//
//  STPAPIClientNetworkBridgeTest.m
//  StripeiOS
//
//  Created by David Estes on 9/23/21.
//  Copyright © 2021 Stripe, Inc. All rights reserved.
//

@import Stripe;
@import XCTest;
@import PassKit;
@import StripeCoreTestUtils;
#import "STPNetworkStubbingTestCase.h"
#import "STPTestingAPIClient.h"
#import "STPFixtures.h"

@interface StripeAPIBridgeNetworkTest : XCTestCase

@property (nonatomic) STPAPIClient *client;

@end

// These are a little redundant with the existing
// API tests, but it's a good way to test that the
// bridge works correctly.
@implementation StripeAPIBridgeNetworkTest

- (void)setUp {
    self.client = [[STPAPIClient alloc] initWithPublishableKey:STPTestingDefaultPublishableKey];
    [super setUp];
}

// MARK: Bank Account
- (void)testCreateTokenWithBankAccount {
    XCTestExpectation *exp = [self expectationWithDescription:@"Request complete"];
    STPBankAccountParams *params = [[STPBankAccountParams alloc] init];
    params.accountNumber = @"000123456789";
    params.routingNumber = @"110000000";
    params.country = @"US";
    
    [self.client createTokenWithBankAccount:params completion:^(STPToken *token, NSError *error) {
        XCTAssertNotNil(token);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: PII

- (void)testCreateTokenWithPII {
    XCTestExpectation *exp = [self expectationWithDescription:@"Create token"];
    
    [self.client createTokenWithPersonalIDNumber:@"123456789" completion:^(STPToken *token, NSError *error) {
        XCTAssertNotNil(token);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

- (void)testCreateTokenWithSSNLast4 {
    XCTestExpectation *exp = [self expectationWithDescription:@"Create SSN"];
    
    [self.client createTokenWithSSNLast4:@"1234" completion:^(STPToken *token, NSError *error) {
        XCTAssertNotNil(token);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Connect Accounts

- (void)testCreateConnectAccount {
    XCTestExpectation *exp = [self expectationWithDescription:@"Create connect account"];
    STPConnectAccountCompanyParams *companyParams = [[STPConnectAccountCompanyParams alloc] init];
    companyParams.name = @"Company";
    STPConnectAccountParams *params = [[STPConnectAccountParams alloc] initWithCompany:companyParams];
    [self.client createTokenWithConnectAccount:params completion:^(STPToken *token, NSError *error) {
        XCTAssertNotNil(token);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Upload

- (void)testUploadFile {
    XCTestExpectation *exp = [self expectationWithDescription:@"Upload file"];
    UIImage *image = [UIImage imageNamed:@"stp_test_upload_image.jpeg"
                                inBundle:[NSBundle bundleForClass:self.class]
           compatibleWithTraitCollection:nil];
    
    [self.client uploadImage:image purpose:STPFilePurposeDisputeEvidence completion:^(STPFile *file, NSError *error) {
        XCTAssertNotNil(file);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Credit Cards

- (void)testCardToken {
    XCTestExpectation *exp = [self expectationWithDescription:@"Create card token"];
    STPCardParams *params = [[STPCardParams alloc] init];
    params.number = @"4242424242424242";
    params.expYear = 42;
    params.expMonth = 12;
    params.cvc = @"123";
    
    [self.client createTokenWithCard:params completion:^(STPToken *token, NSError *error) {
        XCTAssertNotNil(token);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

- (void)testCVCUpdate  {
    XCTestExpectation *exp = [self expectationWithDescription:@"CVC Update"];
    
    [self.client createTokenForCVCUpdate:@"123" completion:^(STPToken *token, NSError *error) {
        XCTAssertNotNil(token);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Sources

- (void)testCreateRetrieveAndPollSource  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Upload file"];
    XCTestExpectation *expR = [self expectationWithDescription:@"Retrieve source"];
    XCTestExpectation *expP = [self expectationWithDescription:@"Poll source"];

    STPCardParams *card = [[STPCardParams alloc] init];
    card.number = @"4242424242424242";
    card.expYear = 42;
    card.expMonth = 12;
    card.cvc = @"123";
    
    STPSourceParams *params = [STPSourceParams cardParamsWithCard:card];
    
    [self.client createSourceWithParams:params completion:^(STPSource *source, NSError *error) {
        XCTAssertNotNil(source);
        XCTAssertNil(error);
        [exp fulfill];
        
        [self.client retrieveSourceWithId:source.stripeID clientSecret:source.clientSecret completion:^(STPSource *source2, NSError *error2) {
            XCTAssertNotNil(source2);
            XCTAssertNil(error2);
            [expR fulfill];
        }];

        [self.client startPollingSourceWithId:source.stripeID clientSecret:source.clientSecret timeout:10 completion:^(STPSource *source2, NSError *error2) {
            XCTAssertNotNil(source2);
            XCTAssertNil(error2);
            [self.client stopPollingSourceWithId:source.stripeID];
            [expP fulfill];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Payment Intents

- (void)testRetrievePaymentIntent  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Fetch"];
    XCTestExpectation *exp2 = [self expectationWithDescription:@"Fetch with expansion"];
    
    STPTestingAPIClient *testClient = [[STPTestingAPIClient alloc] init];
    [testClient createPaymentIntentWithParams:nil completion:^(NSString *clientSecret, NSError *error) {
        XCTAssertNil(error);
        
        [self.client retrievePaymentIntentWithClientSecret:clientSecret completion:^(STPPaymentIntent *pi, NSError *error2) {
            XCTAssertNotNil(pi);
            XCTAssertNil(error2);
            [exp fulfill];
        }];
        
        [self.client retrievePaymentIntentWithClientSecret:clientSecret expand:@[@"metadata"] completion:^(STPPaymentIntent *pi, NSError *error2) {
            XCTAssertNotNil(pi);
            XCTAssertNil(error2);
            [exp2 fulfill];
        }];
    }];
     
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

- (void)testConfirmPaymentIntent  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Confirm"];
    XCTestExpectation *exp2 = [self expectationWithDescription:@"Confirm with expansion"];
    STPTestingAPIClient *testClient = [[STPTestingAPIClient alloc] init];
    
    STPPaymentMethodCardParams *card = [[STPPaymentMethodCardParams alloc] init];
    card.number = @"4242424242424242";
    card.expYear = @42;
    card.expMonth = @12;
    card.cvc = @"123";
    
    [testClient createPaymentIntentWithParams:nil completion:^(NSString *clientSecret, NSError *error) {
        XCTAssertNil(error);
        
        STPPaymentIntentParams *params = [[STPPaymentIntentParams alloc] initWithClientSecret:clientSecret];
        params.paymentMethodParams = [STPPaymentMethodParams paramsWithCard:card billingDetails:nil metadata:nil];
        
        [self.client confirmPaymentIntentWithParams:params completion:^(STPPaymentIntent *pi, NSError *error2) {
            XCTAssertNotNil(pi);
            XCTAssertNil(error2);
            [exp fulfill];
        }];
    }];
    
    [testClient createPaymentIntentWithParams:nil completion:^(NSString *clientSecret, NSError *error) {
        XCTAssertNil(error);
        
        STPPaymentIntentParams *params = [[STPPaymentIntentParams alloc] initWithClientSecret:clientSecret];
        params.paymentMethodParams = [STPPaymentMethodParams paramsWithCard:card billingDetails:nil metadata:nil];
        
        [self.client confirmPaymentIntentWithParams:params completion:^(STPPaymentIntent *pi, NSError *error2) {
            XCTAssertNotNil(pi);
            XCTAssertNil(error2);
            [exp2 fulfill];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Setup Intents

- (void)testRetrieveSetupIntent  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Fetch"];
    
    STPTestingAPIClient *testClient = [[STPTestingAPIClient alloc] init];
    [testClient createSetupIntentWithParams:nil completion:^(NSString *clientSecret, NSError *error) {
        XCTAssertNil(error);
        
        [self.client retrieveSetupIntentWithClientSecret:clientSecret completion:^(STPSetupIntent *si, NSError *error2) {
            XCTAssertNotNil(si);
            XCTAssertNil(error2);
            [exp fulfill];
        }];
    }];
     
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

- (void)testConfirmSetupIntent  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Confirm"];
    STPTestingAPIClient *testClient = [[STPTestingAPIClient alloc] init];
    
    STPPaymentMethodCardParams *card = [[STPPaymentMethodCardParams alloc] init];
    card.number = @"4242424242424242";
    card.expYear = @42;
    card.expMonth = @12;
    card.cvc = @"123";
    
    [testClient createSetupIntentWithParams:nil completion:^(NSString *clientSecret, NSError *error) {
        XCTAssertNil(error);
        
        STPSetupIntentConfirmParams *params = [[STPSetupIntentConfirmParams alloc] initWithClientSecret:clientSecret];
        params.paymentMethodParams = [STPPaymentMethodParams paramsWithCard:card billingDetails:nil metadata:nil];
        
        [self.client confirmSetupIntentWithParams:params completion:^(STPSetupIntent *si, NSError *error2) {
            XCTAssertNotNil(si);
            XCTAssertNil(error2);
            [exp fulfill];
        }];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: Payment Methods

- (void)testCreatePaymentMethod  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Create PaymentMethod"];

    STPPaymentMethodCardParams *card = [[STPPaymentMethodCardParams alloc] init];
    card.number = @"4242424242424242";
    card.expYear = @42;
    card.expMonth = @12;
    card.cvc = @"123";
    
    STPPaymentMethodParams *params = [[STPPaymentMethodParams alloc] initWithCard:card billingDetails:nil metadata:nil];
    
    [self.client createPaymentMethodWithParams:params completion:^(STPPaymentMethod *pm, NSError *error) {
        XCTAssertNotNil(pm);
        XCTAssertNil(error);
        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}


// MARK: Radar

- (void)testCreateRadarSession  {
    XCTestExpectation *exp = [self expectationWithDescription:@"Create session"];

    [self.client createRadarSessionWithCompletion:^(STPRadarSession *session, NSError *error) {
        XCTAssertNotNil(session);
        XCTAssertNil(error);
        [exp fulfill];
    }];

    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

// MARK: ApplePay

- (void)testCreateApplePayToken  {
    XCTestExpectation *exp = [self expectationWithDescription:@"CreateToken"];
    XCTestExpectation *exp2 = [self expectationWithDescription:@"CreateSource"];
    XCTestExpectation *exp3 = [self expectationWithDescription:@"CreatePM"];
    PKPayment *payment = [STPFixtures applePayPayment];
    [self.client createTokenWithPayment:payment completion:^(STPToken *token, NSError *error) {
        // The certificate used to sign our fake Apple Pay test payment is invalid, which makes sense.
        // Expect an error.
        XCTAssertNil(token);
        XCTAssertNotNil(error);
        [exp fulfill];
    }];
    
    [self.client createSourceWithPayment:payment completion:^(STPSource *source, NSError *error) {
        XCTAssertNil(source);
        XCTAssertNotNil(error);
        [exp2 fulfill];
    }];
    
    [self.client createPaymentMethodWithPayment:payment completion:^(STPPaymentMethod *pm, NSError *error) {
        XCTAssertNil(pm);
        XCTAssertNotNil(error);
        [exp3 fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

- (void)testPKPaymentError {
    XCTestExpectation *exp = [self expectationWithDescription:@"Upload file"];
    STPCardParams *params = [[STPCardParams alloc] init];
    params.number = @"4242424242424242";
    params.expYear = 20;
    params.expMonth = 12;
    params.cvc = @"123";
    
    [self.client createTokenWithCard:params completion:^(STPToken *token, NSError *error) {
        XCTAssertNil(token);
        XCTAssertNotNil(error);
        XCTAssertNotNil([STPAPIClient pkPaymentErrorForStripeError:error]);

        [exp fulfill];
    }];
    
    [self waitForExpectationsWithTimeout:TestConstants.STPTestingNetworkRequestTimeout handler:nil];
}

@end

File: Stripe/StripeiOSTests/STPAPIClientNetworkBridgeTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPAPIClientNetworkBridgeTest.m
//  StripeiOS
//
//  Created by David Estes on 9/23/21.
//  Copyright © 2021 Stripe, Inc. All rights reserved.
//

import PassKit
import Stripe
@testable@_spi(STP) import StripeCore
import StripeCoreTestUtils
@_spi(STP) import StripePayments
import StripePaymentsTestUtils
import XCTest

class StripeAPIBridgeNetworkTest: STPNetworkStubbingTestCase {
    var client: STPAPIClient!

    override func setUp() {
        super.setUp()
        client = STPAPIClient(publishableKey: STPTestingDefaultPublishableKey)
    }

    // MARK: Bank Account
    func testCreateTokenWithBankAccount() {
        let exp = expectation(description: "Request complete")
        let params = STPBankAccountParams()
        params.accountNumber = "000123456789"
        params.routingNumber = "110000000"
        params.country = "US"

        client?.createToken(withBankAccount: params) { token, error in
            XCTAssertNotNil(token)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: PII

    func testCreateTokenWithPII() {
        let exp = expectation(description: "Create token")

        client?.createToken(withPersonalIDNumber: "123456789") { token, error in
            XCTAssertNotNil(token)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testCreateTokenWithSSNLast4() {
        let exp = expectation(description: "Create SSN")

        client?.createToken(withSSNLast4: "1234") { token, error in
            XCTAssertNotNil(token)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Connect Accounts

    func testCreateConnectAccount() {
        let exp = expectation(description: "Create connect account")
        let companyParams = STPConnectAccountCompanyParams()
        companyParams.name = "Company"
        let params = STPConnectAccountParams(company: companyParams)
        client?.createToken(withConnectAccount: params) { token, error in
            XCTAssertNotNil(token)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Upload

    func testUploadFile() {
        let exp = expectation(description: "Upload file")
        let image = UIImage(
            named: "stp_test_upload_image.jpeg",
            in: Bundle(for: StripeAPIBridgeNetworkTest.self),
            compatibleWith: nil)!

        client?.uploadImage(image, purpose: .disputeEvidence) { file, error in
            XCTAssertNotNil(file)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Credit Cards

    func testCardToken() {
        let exp = expectation(description: "Create card token")
        let params = STPCardParams()
        params.number = "4242424242424242"
        params.expYear = 42
        params.expMonth = 12
        params.cvc = "123"

        client?.createToken(withCard: params) { token, error in
            XCTAssertNotNil(token)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testCVCUpdate() {
        let exp = expectation(description: "CVC Update")

        client?.createToken(forCVCUpdate: "123") { token, error in
            XCTAssertNotNil(token)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Sources

    func testCreateRetrieveAndPollSource() {
        let exp = expectation(description: "Upload file")
        let expR = expectation(description: "Retrieve source")
        let expP = expectation(description: "Poll source")

        let card = STPCardParams()
        card.number = "4242424242424242"
        card.expYear = 42
        card.expMonth = 12
        card.cvc = "123"

        let params = STPSourceParams.cardParams(withCard: card)

        client.createSource(with: params) { [self] source, error in
            guard let source = source else {
                XCTFail()
                return
            }
            XCTAssertNil(error)
            exp.fulfill()

            client?.retrieveSource(withId: source.stripeID, clientSecret: source.clientSecret!) { source2, error2 in
                XCTAssertNotNil(source2)
                XCTAssertNil(error2)
                expR.fulfill()
            }

            client?.startPollingSource(withId: source.stripeID, clientSecret: source.clientSecret!, timeout: 10) { [self] source2, error2 in
                XCTAssertNotNil(source2)
                XCTAssertNil(error2)
                client?.stopPollingSource(withId: source.stripeID)
                expP.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Payment Intents

    func testRetrievePaymentIntent() {
        let exp = expectation(description: "Fetch")
        let exp2 = expectation(description: "Fetch with expansion")

        let testClient = STPTestingAPIClient.shared()
        testClient.createPaymentIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            client?.retrievePaymentIntent(withClientSecret: clientSecret!) { pi, error2 in
                XCTAssertNotNil(pi)
                XCTAssertNil(error2)
                exp.fulfill()
            }

            client?.retrievePaymentIntent(withClientSecret: clientSecret!, expand: ["metadata"]) { pi, error2 in
                XCTAssertNotNil(pi)
                XCTAssertNil(error2)
                exp2.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testConfirmPaymentIntent() {
        let exp = expectation(description: "Confirm")
        let exp2 = expectation(description: "Confirm with expansion")
        let testClient = STPTestingAPIClient.shared()

        let card = STPPaymentMethodCardParams()
        card.number = "4242424242424242"
        card.expYear = NSNumber(value: 42)
        card.expMonth = NSNumber(value: 12)
        card.cvc = "123"

        testClient.createPaymentIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            let params = STPPaymentIntentParams(clientSecret: clientSecret!)
            params.paymentMethodParams = STPPaymentMethodParams(card: card, billingDetails: nil, metadata: nil)

            client?.confirmPaymentIntent(with: params) { pi, error2 in
                XCTAssertNotNil(pi)
                XCTAssertNil(error2)
                exp.fulfill()
            }
        }

        testClient.createPaymentIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            let params = STPPaymentIntentParams(clientSecret: clientSecret!)
            params.paymentMethodParams = STPPaymentMethodParams(card: card, billingDetails: nil, metadata: nil)

            client?.confirmPaymentIntent(with: params) { pi, error2 in
                XCTAssertNotNil(pi)
                XCTAssertNil(error2)
                exp2.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testRefreshPaymentIntent() {
        let exp = expectation(description: "Refresh")

        let testClient = STPTestingAPIClient.shared()
        testClient.createPaymentIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            client?.refreshPaymentIntent(withClientSecret: clientSecret!) { pi, error2 in
                XCTAssertNotNil(pi)
                XCTAssertNil(error2)
                exp.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Setup Intents

    func testRetrieveSetupIntent() {
        let exp = expectation(description: "Fetch")

        let testClient = STPTestingAPIClient.shared()
        testClient.createSetupIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            client?.retrieveSetupIntent(withClientSecret: clientSecret!) { si, error2 in
                XCTAssertNotNil(si)
                XCTAssertNil(error2)
                exp.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testConfirmSetupIntent() {
        let exp = expectation(description: "Confirm")
        let testClient = STPTestingAPIClient.shared()

        let card = STPPaymentMethodCardParams()
        card.number = "4242424242424242"
        card.expYear = NSNumber(value: 42)
        card.expMonth = NSNumber(value: 12)
        card.cvc = "123"

        testClient.createSetupIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            let params = STPSetupIntentConfirmParams(clientSecret: clientSecret!)
            params.paymentMethodParams = STPPaymentMethodParams(card: card, billingDetails: nil, metadata: nil)

            client?.confirmSetupIntent(with: params) { si, error2 in
                XCTAssertNotNil(si)
                XCTAssertNil(error2)
                exp.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testRefreshSetupIntent() {
        let exp = expectation(description: "Refresh")

        let testClient = STPTestingAPIClient.shared()
        testClient.createSetupIntent(withParams: nil) { [self] clientSecret, error in
            XCTAssertNil(error)

            client?.refreshSetupIntent(withClientSecret: clientSecret!) { si, error2 in
                XCTAssertNotNil(si)
                XCTAssertNil(error2)
                exp.fulfill()
            }
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Payment Methods

    func testCreatePaymentMethod() {
        let exp = expectation(description: "Create PaymentMethod")

        let card = STPPaymentMethodCardParams()
        card.number = "4242424242424242"
        card.expYear = NSNumber(value: 42)
        card.expMonth = NSNumber(value: 12)
        card.cvc = "123"

        let params = STPPaymentMethodParams(card: card, billingDetails: nil, metadata: nil)

        client?.createPaymentMethod(with: params) { pm, error in
            XCTAssertNotNil(pm)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: Radar

    func testCreateRadarSession() {
        let exp = expectation(description: "Create session")

        // Set fake SID/MUID to make this test replicable
        FraudDetectionData.shared.sid = "123"
        FraudDetectionData.shared.muid = "123"
        FraudDetectionData.shared.sidCreationDate = Date()

        client?.createRadarSession { session, error in
            XCTAssertNotNil(session)
            XCTAssertNil(error)
            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    // MARK: ApplePay

    func testCreateApplePayToken() {
        let exp = expectation(description: "CreateToken")
        let exp2 = expectation(description: "CreateSource")
        let exp3 = expectation(description: "CreatePM")
        let payment = STPFixtures.applePayPayment()
        client?.createToken(with: payment) { token, error in
            // The certificate used to sign our fake Apple Pay test payment is invalid, which makes sense.
            // Expect an error.
            XCTAssertNil(token)
            XCTAssertNotNil(error)
            exp.fulfill()
        }

        client?.createSource(with: payment) { source, error in
            XCTAssertNil(source)
            XCTAssertNotNil(error)
            exp2.fulfill()
        }

        client?.createPaymentMethod(with: payment) { pm, error in
            XCTAssertNil(pm)
            XCTAssertNotNil(error)
            exp3.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }

    func testPKPaymentError() {
        let exp = expectation(description: "Upload file")
        let params = STPCardParams()
        params.number = "4242424242424242"
        params.expYear = 20
        params.expMonth = 12
        params.cvc = "123"

        client?.createToken(withCard: params) { token, error in
            XCTAssertNil(token)
            XCTAssertNotNil(error)
            XCTAssertNotNil(STPAPIClient.pkPaymentError(forStripeError: error))

            exp.fulfill()
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }
}

// These are a little redundant with the existing
// API tests, but it's a good way to test that the
// bridge works correctly.
